src/core/mac/sub_mac.cpp

```diff
@@ -591,10 +611,30 @@ void SubMac::HandleTransmitDone(TxFrame &aFrame, RxFrame *aAckFrame, Error aErro
     {
         mTransmitRetries++;
         aFrame.SetIsARetransmission(true);
+
+#if OPENTHREAD_CONFIG_MAC_ADD_DELAY_ON_NO_ACK_ERROR_BEFORE_RETRY
+        if (aError == kErrorNoAck)
+        {
+            SetState(kStateDelayBeforeRetx);
+            StartTimerForBackoff(mRetxDelayBackOffExponent);
+            mRetxDelayBackOffExponent = OT_MIN(mRetxDelayBackOffExponent + 1, kRetxDelayMaxBackoffExponent);
+            ExitNow();
+        }
+#endif
+
         StartCsmaBackoff();
         ExitNow();
     }
 
+#if OPENTHREAD_CONFIG_MAC_ADD_DELAY_ON_NO_ACK_ERROR_BEFORE_RETRY
+#if OPENTHREAD_CONFIG_MAC_MAINTAIN_RETX_DELAY_BACKOFF_EXPONENT
+    if ((aError == kErrorNone) && aFrame.GetAckRequest())
+#endif
+    {
+        mRetxDelayBackOffExponent = kRetxDelayMinBackoffExponent;
+    }
+#endif
+
     SetState(kStateReceive);
```

Comment:
One minor clarification here about this approach, if we do not reset the backoff exponent  (OPENTHREAD_CONFIG_MAC_MAINTAIN_RETX_DELAY_BACKOFF_EXPONENT is not set) , there could be multiple receive operations or Indirect transmissions between the next direct transmission.  The time at which the exponent would get applied might be far off and channel conditions could be different right?

Reply:
> there could be multiple receive operations or Indirect transmissions between the next direct transmission. The time at which the exponent would get applied might be far off and channel conditions could be different right

The random wait time (once picked) will be fixed and cannot be pushed back. For example if we are at `BE=5` and pick `6 msec` as wait time, we will send the frame after `6 ms`. We can potentially receive frames during the backoff wait interval (which will be passed to next layer `Mac` but they cannot really impact the current wait time or change the current tx operation). `SubMac` layer deals with a single ongoing tx at a time and the notions of direct vs indirect come into picture at `Mac` or next layers.